### PR TITLE
Fix Validation Step breaking Multi-GPU Training

### PR DESCRIPTION
**Pull request by RuntimeRacer**
I propose this PR to fix issues I encountered with Multi-GPU training.

Based on my own testing and research I found out:

- DDP Models keep themselves in sync between threads without using join context
- Join context implementation is not properly handled in the current implementation, according to docs (https://pytorch.org/tutorials/advanced/generic_join.html) this requires implementing the Joinable interface in the model class itself, plus additional hooks. However, the current code uses a `model.join` param which is actually a null pointer and therefore no different to nullcontext, plus does not keep the individual training threads in sync.
- If proper handling of the training threads is applied manually, and dataloaders support DDP approach with splitting of batches across world size (Which lhotse dataloaders support natively), the implementation of the Joinable interface is not required.

Therefore this PR applies the following changes:
- Removes broken join handling of training threads
- Explicit Synchronization of threads when performing the following operations in the main thread:
  - Saving of a model checkpoint
  - Updating the Averaged model
  - Computation of the validation loss (I found that there is no difference in validation loss result if performed by all threads at the same time, so it can be handled by one alone)
- Omitting the OOM check at training start if epoch > 1 or flag is set to omit it.

As intermediate Results, with these changes I was able to see the following results so far:
- No crashes / hangs on Validation Loss computation
- While training a whole epoch of 376k steps on a single GPU needed 28 hours to complete, I was able to train another epoch on 4 GPUs, which needed 95k steps and 20 hours to complete.

### CODE REVIEW

**Comment by lifeiteng** [comment]
@chenjiasheng  Can you take a look?

File: valle/bin/trainer.py [hunk 1]
```diff
@@ -1002,7 +1017,7 @@ def run(rank, world_size, args):
     )
     valid_dl = dataset.valid_dataloaders(valid_cuts)
 
-    if True:
+    if params.oom_check and params.start_epoch == 1:
```
Comment (lifeiteng) [inline valle/bin/trainer.py:1020]: `if params.oom_check` let's check oom when training retart

**Comment by lifeiteng** [comment]
@RuntimeRacer thank you for contributing, can you update according to comments?

File: valle/bin/trainer.py (hunk 1 above)
Comment (pawel-polyai) [reply valle/bin/trainer.py:1020]: +1

**Review by pawel-polyai — APPROVED** [review]
Lgtm. I have run a small test check with 8 GPUs, no issues during validation pass. The torch.distributed.barrier solves the issue. 
Screenshot 2023-05-11 at 12 23 16 PM

File: valle/bin/trainer.py (hunk 1 above)
Comment (RuntimeRacer) [reply valle/bin/trainer.py]: Updated the PR accoringly.

**Comment by RuntimeRacer** [comment]
> @RuntimeRacer thank you for contributing, can you update according to comments?

@lifeiteng yes I updated it.

**Review by lifeiteng — APPROVED** [review]

**Comment by chenjiasheng** [comment]
Sorry for the late.
New added `model.train()`, 'model.valid()' and `if rank == 0` checks are redundant because the following functions `compute_loss()`, `load_xxx_weights()` has that check inside the functions. However it is OK to add these for explicity.

`model.join` context actually works, as I have confirmed in #95. The reason why it didn't fix our validate hang issue is due to unequal length dataloaders of ranks. the `model.join` context works by having the ranks that run out of data first emulate a void `forward` and `backward` in the context's `__exit__` method, while our validation hang issue is because of some ranks are doing validation whereas others are doing training. 
Of corse putting validation on rank0 solely can solve this issue, but I personally don't like this solution.
I use a similar approach as lhotse as @RuntimeRacer mentioned, I cut dataloaders to the shortest rank length before an epoch.

In a word, I think this PR is fine.
Thank you so much @RuntimeRacer
